src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -2387,6 +2387,19 @@ QD_EXPORT void qd_dispatch_delete_tcp_listener(qd_dispatch_t *qd, void *impl)
             listener->common.vflow = 0;
         }
         //
+        // Initiate termination of existing connections
+        //
+        if (tcp_context->qd->terminate_tcp_conns) {
+            sys_mutex_lock(&listener->lock);
+            qd_tcp_connection_t *conn = DEQ_HEAD(listener->connections);
```

Comment:
Each member of `listener->connections` must be removed so that at the end of this loop `listener->connections` will be empty

Something like this - 

```
sys_mutex_lock(&listener->lock);
 qd_tcp_connection_t *conn = DEQ_HEAD(listener->connections);

 while (conn) {
     qd_tcp_connection_t *next_conn = DEQ_NEXT(conn);
     DEQ_REMOVE_HEAD(listener->connections);
     qdr_core_close_connection(conn->core_conn);
     conn = next_conn;
 }
 sys_mutex_unlock(&listener->lock);
```

Reply:
Please ignore this comment

Reply:
```
/home/runner/work/skupper-router/skupper-router/skupper-router/src/router_core/connections.c:338:5: runtime error: member access within null pointer of type 'struct qdr_connection_t'
```
The  error points to this line: https://github.com/gabordozsa/skupper-router/blob/0f07033f08a7a98d5ade343c13763271920dc5b5/src/router_core/connections.c#L338
It seems that `conn->core_conn` is null pointer when `qdr_core_close_connection(conn->core_conn)` is called in the connections traversing loop.   That may happen if` close_connection_XSIDE_IO()` has already called for this `conn`.  Does that make sense? If so, I do not see how removing the this `conn` from the `connections` list would fix this.

Reply:
@ganeshmurthy  I pushed two new commits as an attempt to fix this error.

1.  I take the` conn->activation_lock` and check if `conn->raw_opened` is set before calling `qdr_core_close_connection(conn->core_conn)`.  The reasoning here is that `close_connection_XSIDE_IO()` can zero  `conn->core_conn`  but it also clears  `conn->raw_opened` before. If the ` conn->activation_lock` is taken then `close_connection_XSIDE_IO()` cannot change the value of `conn->core_conn` while we are calling `qdr_core_close_connection(conn->core_conn)`.
2. I  also check if `conn->core_conn` is not zero before calling `qdr_core_close_connection(conn->core_conn)`. This is because `conn->core_conn` is initialised when the the link is created in ` link_setup_[L,C]SIDE_IO ()` but the `conn->raw_opened` is set before that at the PN_RAW_CONNECTION_CONNECTED event. 

Does that make sense? I think that still more work is necessary for this PR because the TCP connection is not terminated if case (2) happens (i.e. tcpListener is deleted which has a connection with 'raw_opened' set but 'core_conn' is still zero because the link is not setup yet).

Reply:
Yes, that makes sense @gabordozsa 

I would like to state the problem in simple terms - 
The `PN_RAW_CONNECTION_CONNECTED` event or the `PN_RAW_CONNECTION_DISCONNECTED` event could come upon any of the connections in `listener->connections` while the listener is being deleted.

Case 1: When a connection receives a `PN_RAW_CONNECTION_DISCONNECTED` event at the same time of the listener deletion (listener deletion call runs on a management thread and `PN_RAW_CONNECTION_DISCONNECTED` event comes in via an i/o thread),  your if statement - 
`if (!!conn->core_conn && IS_ATOMIC_FLAG_SET(&conn->raw_opened))` will make sure that we don't call `qdr_core_close_connection(conn->core_conn)` on an already closed connection. So, this case is covered.

Case 2: A connection receives a `PN_RAW_CONNECTION_CONNECTED` event at the same time of the listener deletion. The listener delete function might miss deleting the connection because `conn->core_conn` is not set yet. This will lead to a leaking `qd_tcp_connection_t` and `qdr_connection_t`.

Reply:
@ganeshmurthy  I added a new `pending_close` flag to `qd_tcp_connection_t` as per your suggestion. The `connection_run_[L,C]SIDE_IO()` functions closing the connection if this flag is set. I also set the existing `closing` flag in `tcp_listener` and `tcp_connector` in order to set the `pending_close` flag for connections which are not in the `connections` list yet (e.g. when `on_accept()` created the new `conn` but has not called `DEQ_INSERT_TAIL(conn)` yet). I think the `closing` flag was unused before.